File: main/course-schedule/course-schedule.py
import enum

class Solution:
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        if numCourses < 0:
            raise ValueError("can't have negatively many courses")

        adj = Solution._buildAdjacencyList(numCourses, prerequisites)
        return not Solution._hasCycle(adj)

    @staticmethod
    def _buildAdjacencyList(numCourses, prerequisites):
        adj = [[] for _ in range(numCourses)]

        for dest, src in prerequisites:
            adj[src].append(dest)

        return adj

    @staticmethod
    def _hasCycle(adj):
        vis = [Solution._Color.WHITE] * len(adj)

        def dfs(src):
            if vis[src] is Solution._Color.WHITE:
                vis[src] = Solution._Color.GRAY
                if any(dfs(dest) for dest in adj[src]):
                    return True
                vis[src] = Solution._Color.BLACK
                return False
            elif vis[src] is Solution._Color.GRAY:
                return True
            elif vis[src] is Solution._Color.BLACK:
                return False
            else:
                raise AssertionError('invalid visitation state')

        return any(dfs(src) for src in range(len(vis)))

    @enum.unique
    class _Color(enum.Enum):
        WHITE = 0
        GRAY = 1
        BLACK = 2

File: main/course-schedule/course_schedule.rb
def adjacency_list(vertex_count, edges)
  adj = Array.new(vertex_count) { [] }
  edges.each { |src, dest| adj[src] << dest }
  adj
end

def cycle?(adj)
  vis = Array.new(adj.size, :white)

  has_cycle_from = lambda do |src|
    case vis[src]
    when :white
      vis[src] = :gray
      return true if adj[src].any?(&has_cycle_from)
      vis[src] = :black
      false
    when :gray
      true
    when :black
      false
    else
      raise 'Internal error: invalid visitation state!'
    end
  end

  adj.size.times.any?(&has_cycle_from)
end

# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  !cycle?(adjacency_list(num_courses, prerequisites))
end
